pool/minerid.go

```diff
@@ -0,0 +1,59 @@
+// Copyright (c) 2020-2023 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package pool
+
+import (
+	"fmt"
+	"regexp"
+
+	errs "github.com/decred/dcrpool/errors"
+)
+
+// newUserAgentRE returns a compiled regular expression that matches a user
+// agent with the provided client name, major version, and minor version as well
+// as any patch, pre-release, and build metadata suffix that are valid per the
+// semantic versioning 2.0.0 spec.
+//
+// For reference, user agents are expected to be of the form "name/version"
+// where the name is a string and the version follows the semantic versioning
+// 2.0.0 spec.
+func newUserAgentRE(clientName string, clientMajor, clientMinor uint32) *regexp.Regexp {
+	// semverBuildAndMetadataSuffixRE is a regular expression to match the
+	// optional pre-release and build metadata portions of a semantic version
+	// 2.0 string.
+	const semverBuildAndMetadataSuffixRE = `(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-]` +
+		`[0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?` +
+		`(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?`
+
+	return regexp.MustCompile(fmt.Sprintf(`^%s\/%d\.%d\.(0|[1-9]\d*)%s$`,
+		clientName, clientMajor, clientMinor, semverBuildAndMetadataSuffixRE))
+}
+
+var (
+	// These regular expressions are used to identify the expected mining
+	// clients by the user agents in their mining.subscribe requests.
+	cpuRE = newUserAgentRE("cpuminer", 1, 0)
+	nhRE  = newUserAgentRE("NiceHash", 1, 0)
+
+	// miningClients maps regular expressions to the supported mining client IDs
+	// for all user agents that match the regular expression.
+	miningClients = map[*regexp.Regexp][]string{
```

Comment:
```suggestion
	miningClients = map[*regexp.Regexp]string{
```

I think this change would serve to simplify things quite a bit. Maybe there is a reason for it, but I don't see why a single ID would need to be mapping to multiple miner types. We can add support for that later if it ever becomes necessary.

Reply:
I left it that way because there is infrastructure for upgrading based on the hash rate.  There is nothing forcing two different devices to use a different user agent, and as a result, you can have multiple devices with wildly different hashrates identifying themselves the same.

There are tests which cover the upgrading as well.

Reply:
As an aside, I think that ideally the pool really generally shouldn't even be assigning hash rates and instead should figure it out dynamically (as the client upgrade code is doing for devices with the same UA).

Then, it could assign some initial hash rates per UA and even better would be for the protocol to allow clients to specify it themselves (which is unfortunately not part of stratum).

However, that is a fair amount of work that doesn't seem worth it given the goal at the moment is primarily getting into shape for BLAKE3 and gominer as opposed to adding new features.